### DISCUSSION TITLE
Contributing a new buildpack

### DISCUSSION THREAD

**Discussion by kevin-ortega** [paketo-buildpacks/feedback#11] [General]
I want to start the conversation about how best to contribute a java buildpack using [open liberty](https://openliberty.io).   This buildpack will be the CN version of the cloud foundry [IBM Webshere Liberty](https://github.com/cloudfoundry/ibm-websphere-liberty-buildpack) buildpack and will include the open source liberty runtime as the default and the commercial WebSphere Liberty as an optional runtime.  Additionally, I want to provided the openj9 v8 and v11 JREs.  

In short, this will be similar to the java buildpack but replace tomcat with open liberty and bellsoft-liberica with openj9.  

I have been a contributor to the cloud foundry ibm websphere buildpack for a few years.  

The goal is to contribute the buildpack by the end of the 2nd QTR 2021.  

I already have a minimal working buildpack that provides either of the liberty runtimes and either openj9 v8 or v11 when the detect finds a war. .  It is based on the apache-tomcat buildpack.  

I understand I'll need to submit an RFC.  

What can I do to ensure the RFC will be accepted and the buildpack can be contributed and available by end of this quarter?   

Thank you.

**Comment by dmikusa** [comment]
Hi @kevin-ortega thanks for reaching out. Happy to give you some suggestions here.

1. The Paketo buildpacks & cloud-native buildpacks in general are a lot more composable than the previous buildpacks, so I hope you'll find this process is easier than before. Anything new would want to follow this model making buildpacks that do one thing and composing them together.

2. What we have already is an existing "java" buildpack with some opinions about the JVM vendor & app server to use. This is what we're calling a multi-buildpack or meta-buildpack. It basically just references and pulls together in a particular order, other buildpacks. If you look at https://github.com/paketo-buildpacks/java/blob/main/buildpack.toml, you can see what I mean. It's just composing things in a certain order. You could very easily take this and make an "ibm-java" (doesn't have to be this name, but the name for this would typically describe what it does, not what technology it uses) buildpack that is just composed of different stuff.

3. Right now there is no Liberty buildpack for the server itself. If you look at the Tomcat buildpack, you can see an example (it sounds like you're doing this already, which is good). In a nutshell, this buildpack will want to download Liberty, install it and drop any configurations required into the layers it generates. It should be small and do one thing so that it's potentially composable in other multi/meta buildpacks. The name for this buildpack would describe what is being installed, so this is where something like `ibm-websphere-liberty` would make sense as a name.

4. Good news is that we already have an [OpenJ9 buildpack](https://github.com/paketo-buildpacks/eclipse-openj9), so you can make use of that. If you want to contribute other JVMs, you can do so. We have a [libjvm](https://github.com/paketo-buildpacks/libjvm) project for making this as simple as possible. A new JVM buildpack should definitely use libjvm.

So in summary, it sounds like you'd want to propose two buildpacks. One a multi/meta buildpack and one that installs/configure Liberty. 

I'm not sure we could guarantee any timelines, but I would encourage you to join our next [Paketo Working Group meeting](https://github.com/paketo-buildpacks/community/#working-group-meetings) & discuss more.

**Comment by krismarc** [comment]
Hi @kevin-ortega, it's been a while. Hope you are all good.

Wondering if you are aware that Cloud Foundry has introduced CNBs as one of possible deployment options.
https://docs.cloudfoundry.org/buildpacks/cnb/

I'm yet using your old good liberty buildpack and replacing it with this CNB could be an option for migration.

1) Do you know if anyone already tried Liberty CNB this way?
2) Is IBM anyhow interested to officially support it with Cloud Foundry?

My very first attempt went successfully against staging process and just failed to get the app started. Probably I need to meet some more criteria. All 'pack' operations seems to be wrapped into cf push command. 

Best regards,
K.M.

**Comment by kevin-ortega** [comment]
Hi @krismarc, good to hear from you.  

I am aware the CF was looking into supporting CNBs but I was not aware it has been introduced.  I am not aware of anyone that has actually tried using the CN liberty buildpack in CF.  It is not likely IBM would officially support the buildpack in CF simply because we don't have access to a CF environment to be able to test with.  I will approach our product managers and tech leads and get back to you but my feeling is they would say it would not be supported.

**Comment by kevin-ortega** [comment]
@krismarc The response I got back from our leads is we would probably support the buildpack running in CF.

**Reply by krismarc** [reply]
@kevin-ortega, That would be great. I'll try to contribute as a tester 😄

**Reply by krismarc** [reply]
Hi @kevin-ortega  👋 It's been about a year since our last exchange, so I thought I'd check in. Has there been any progress or update regarding official support for the Liberty CNB on Cloud Foundry?

I'm still interested in this direction and would be happy to help test or provide feedback if there's anything available.

Thanks!

**Reply by kevin-ortega** [reply]
@krismarc Hi, good to hear from you.  So, we're not going to get an official statement saying we would support the Liberty CNB on Cloud Foundry.   However, no one will be denied support and it would be treated like running any container image on any platform.    We (IBM) no longer have access to a CF environment so that makes it a bit challenging.   We won't be doing anything specifically for CF and would rely on CF adding support for CNB.
